### ISSUE TITLE
Feature: a hook to automatically save the best model during training

### ISSUE THREAD

**Issue by xjinai** [facebookresearch/detectron2#2114]
## 🚀 Feature
I'm using train_net.py to train model using custom dataset. It worked well, especially with multiple GPUs. Thank you for your great work!

I set SOLVER.CHECKPOINT_PERIOD (e.g. 1000) to save the model every 1000 iterations. I want Detectron2 to automatically save the best model as the training goes, so that I can use the best model later for inference and evaluation. Currently this feature is missing in Detectron2. 

## Motivation & Examples

I want Detectron2 to save the best model as the training goes, so that I can use the best model later for inference and evaluation. This feature exists in other object detection frameworks, for example, Darknet. I found that it is very useful since the model may overfit with the specified number of iterations. Therefore, the model_final.pth saved by Detectron2 is typically not the best model. If the best model is saved automatically during training, I can read it later and apply it for inference and evaluation.

A possible implementation:
1) Save the model at the first evaluation period to a file named model_best.pth.
2) At every following evaluation period, compare AP value (on the validation set) with the AP value of the last evaluation period. If AP is better, then automatically update the saved best model (model_best.pth). 

## Note

I believe that this feature is relevant to and would be highly appreciated by many users since saving and applying best model is a common practice for object detection.

**Comment by ppwwyyxx** [comment]
Duplicate of #645

**Comment by xjinai** [comment]
Thank you for your response. 

I looked at #645 before. It is not exactly what I want. What I want is saving the best model during training **as the default behavior**, no matter if I run training from train_net.py or tutorial in Colab notebook. Please consider to add this feature. 

Regarding #645, can you please elaborate on how to add the logic "using the return value of do_test" in plain_train_net.py? Does it work with custom dataset?

**Comment by ppwwyyxx** [comment]
I think it is not reasonable to make it automatically a default because 
* there is no definition of "best" model unless users first choose a metric somehow. So it will require a new user input somewhere
* these example training scripts are examples, as written on the top of those files. They are not meant to support every nice features
* in research, using the best model is not a common thing to do (and sometimes not the right thing to do)

`do_test` in `plain_train_net` returns the metrics. So it works for any dataset as long as there are some metrics.

**Comment by xjinai** [comment]
@ppwwyyxx: Implementing a reusable hook that can be manually added by users to a trainer sounds like a good idea. It would be helpful if an example code (for example, to address my use case above) can be provided. Thank you for your consideration. 

I will try to customize plain_train_net.py. For training custom dataset using multi-GPU on a single machine, should I call my custom register_dataset() function at the beginning of main(args) or before calling launch()?

def main(args):
    # register custom dataset here ?
    cfg = setup(args)
...

if __name__ == "__main__":
    args = default_argument_parser().parse_args()
    print("Command Line Args:", args)
    # or register custom dataset here ?
    launch(...)

For people who have the same needs as I described above, one workaround is to set cfg.SOLVER.CHECKPOINT_PERIOD as the same value as cfg.TEST.EVAL_PERIOD, so that you can select the best model after training completes based on the metric you choose. I hope this helps.

**Comment by ppwwyyxx** [comment]
A hook can be implemented roughly like
```
class BestCheckpointer(Hook):
  def __init__(self, metric, use_max=True):
    ...
  def after_step(self):
    metric = self.trainer.storage.get....
    if is_best:
       self.trainer.checkpointer.save("model_best")
```

datasets can be registered in `main()`. before `launch()` is probably also fine.

**Comment by xjinai** [comment]
Thank you, @ppwwyyxx. For train_net.py, I got it working by registering dataset in main(). I will try plain_train_net.py.

**Comment by rivered** [comment]
Would it be possible to implement this for training with the default engine? 

class MyTrainer(DefaultTrainer):
    @classmethod
    def build_evaluator(cls, cfg, dataset_name, output_folder=None):
        if output_folder is None:
            output_folder = os.path.join(cfg.OUTPUT_DIR, "inference")
        return COCOEvaluator(dataset_name, cfg, True, output_folder)

    def build_hooks(self):
        hooks = super().build_hooks()
        hooks[-3] = PeriodicCheckpointer(self.checkpointer, cfg.SOLVER.CHECKPOINT_PERIOD)
        return hooks

and with this logic change the periodicheckpointer to save only if loss decreases?

**Comment by lixiuyu-cumt** [comment]
Hello,I want to know how do you train with multiple GPUs,when I train network by multiple GPUS, it report the following problem:

RuntimeError: Expected to have finished reduction in the prior iteration before starting a new one. This error indicates that your module has parameters that were not used in producing loss. You can enable unused parameter detection by (1) passing the keyword argument `find_unused_parameters=True` to `torch.nn.parallel.DistributedDataParallel`; (2) making sure all `forward` function outputs participate in calculating loss. If you already have done the above two steps, then the distributed data parallel module wasn't able to locate the output tensors in the return value of your module's `forward` function. Please include the loss function and the structure of the return value of `forward` of your module when reporting this issue (e.g. list, dict, iterable).

Have you ever encountered such a situation?

**Comment by ranjeetthakur** [comment]
```
def after_step(self):
        # No way to use **kwargs
        
        ##ONly do this analys when trainer.iter is divisle by checkpoint_epochs
        curr_val = self.trainer.storage.latest().get('bbox/AP50',0)
        
        try:
            _ = self.trainer.storage.history('max_bbox/AP50')
        except:
            self.trainer.storage.put_scalar('max_bbox/AP50',curr_val)
            
        max_val = self.trainer.storage.history('max_bbox/AP50')._data[-1][0]
            
        if curr_val>max_val:
            self.trainer.storage.put_scalar('max_bbox/AP50', curr_val)
            self.step(self.trainer.iter)

```
Modifying after_step function of Periodic Checkpointer hook in the hook.py file worked for me. I am storing best checkpoint based upon 'max_bbox/AP50' key.

**Comment by onepointconsulting** [comment]
I know there is a Hook for best checkpoints. But here is an example which shows how you can implement one hook for this purpose:

```
class BestCheckpointer(HookBase):

    def before_train(self):
        self.best_metric = 0.0
        self.logger = logging.getLogger("detectron2.trainer")
        self.logger.info("######## Running best check pointer")

    def after_step(self):
        if metric_name in self.trainer.storage._history:
            eval_metric, batches = self.trainer.storage.history(metric_name)._data[-1]
            if self.best_metric < eval_metric:
                self.best_metric = eval_metric
                self.logger.info("######## New best metric", self.best_metric, batches)
                self.trainer.checkpointer.save(f"model_best_{eval_metric:.4f}")

class Trainer(DefaultTrainer):
    @classmethod
    def build_evaluator(cls, cfg, dataset_name, output_folder=None):
        return MAPIOUEvaluator(dataset_name)
    
    def build_hooks(self):
        ret = super().build_hooks()
        ret.append(BestCheckpointer())
        return ret
```

**Comment by ppwwyyxx** [comment]
This has been added in fe4d35b0a8b6b37d45162a68f82e73f4bf7b1ba1

**Comment by sjdch** [comment]
@ranjeetthakur @onepointconsulting 你们修改成功了么？我修改后还是不行，请问如何修改？
谢谢回复
